### PR TITLE
[WIP] Gossip protocol implementation for failure detection

### PR DESCRIPTION
**Pull request by ahanapradhan**
### What type of PR is this?

/kind feature

### What does this PR do / why do we need it:
Gossip protocol for failure detection

### Which issue(s) this PR fixes:

Fixes #

### Special notes for your reviewers:

### CODE REVIEW

File: hetu-docs/en/admin/properties.md [hunk 1]
```diff
@@ -922,3 +922,64 @@ helps with cache affinity scheduling.
 > After the configured time elapsed and no response received, then client connection consider that to be failure in submission of request.
 >
 > (Note: this parameter should be configured with higher time when in high load environment)
+
+## Gossip Protocol Configurations for Failure Detection
```
Comment (Nitin-Kashyap) [inline hetu-docs/en/admin/properties.md:926]: All thse should be under the Failure recovery configuration profile

File: hetu-docs/en/admin/properties.md [hunk 2]
```diff
@@ -922,3 +922,64 @@ helps with cache affinity scheduling.
...
+### `failure-detection-protocol`
+
+>- **Type:** String
+>- **Default value:** `heartbeat`
+> 
+> This property defines the type of failure detector in use. Default configuration is `heartbeat` failure detector.
+> Gossip protocol can be enabled by specifying this parameter in `config.properties` file, with the value `gossip`.
+> All nodes (i.e. coordinator as well as workers) in a cluster should have this property specified in their respective `etc/config.properties` file.
+
+### `failure-detector.heartbeat-interval`
```
Comment (Nitin-Kashyap) [inline hetu-docs/en/admin/properties.md:937]: naming needs standardization

File: hetu-docs/en/admin/properties.md [hunk 3]
```diff
@@ -922,3 +922,64 @@ helps with cache affinity scheduling.
...
+>- **Type:** Duration
+>- **Default value:** `500ms` (500 miliseconds)
+> 
+> This is one of the existing configuration properties, which is used by the gossip protocol.
+> This is the interval of gossip between two nodes in the cluster.
+> In gossip protocol, two workers are expected to gossip with higher frequency than the coordinator and a worker.
+> In `config.properties` for the coordinator, this property can be set with a reasonably higher value, such as `5s` (5 seconds).
+> In workers, this property can be left to use the default value.
+> 
+### `worker-worker-refresh-interval`
```
Comment (Nitin-Kashyap) [inline hetu-docs/en/admin/properties.md:948]: Worker-worker / corrdinator-worker sounds improper

File: presto-main/src/main/java/io/prestosql/failuredetector/CoordinatorGossipFailureDetector.java [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+            public int compare(ServiceDescriptor t2, ServiceDescriptor t1)
+            {
+                return t1.getNodeId().compareTo(t1.getNodeId());
+            }
+        });
+        onlineServices.stream().forEach(o -> uris.add(getHttpUri(o)));
+        return uris;
+    }
+
+    private int[] assignLocalGossipGroups(int idx, int clusterSize)
```
Comment (Nitin-Kashyap) [inline presto-main/src/main/java/io/prestosql/failuredetector/CoordinatorGossipFailureDetector.java:137]: Cljuster size may keep changing thereby causing Indices to become unreliable... better to maintain node-id to ordinal-number map for repeated reference sanity

Comment (ahanapradhan) [reply presto-main/src/main/java/io/prestosql/failuredetector/CoordinatorGossipFailureDetector.java:137]: it is always a pair of ints, containing the start and end index. Nodes are taken as a sorted list. start and end index will give howmuch ever nodes are there for the particular gossip group.

File: hetu-docs/en/admin/properties.md (hunk 2 above)
Comment (ahanapradhan) [reply hetu-docs/en/admin/properties.md:937]: this is existing config from heartbeat failure detector.

File: hetu-docs/en/admin/properties.md (hunk 1 above)
Comment (ahanapradhan) [reply hetu-docs/en/admin/properties.md:926]: ok

**Comment by ahanapradhan** [comment]
This PR has been merged through gitee PR: https://gitee.com/openlookeng/hetu-core/pulls/1469

**Pull request closed without merging**